### PR TITLE
ENH: add `binning` function

### PR DESCRIPTION
**Pull request by kohr-h**
I didn't add any unit tests since the doctests pretty much covers it. Although it might be useful to look at corner cases (`arr.ndim == 0`, `0 in arr.shape`, `bin_size == 1`)... I'll add those.

**TODO**

- [x] Unit test

### CODE REVIEW

**Comment by adler-j** [comment]
Sorry for the non-existent review on this. Do you need one?

**Comment by kohr-h** [comment]
Revived this old PR. Still works, now also with unit tests. Short review?

File: odl/test/util/numerics_test.py [hunk 1]
```diff
@@ -482,5 +573,48 @@ def test_resize_array_raise():
...
+    assert all_almost_equal(out_arr, true_out_arr)
+    assert out_arr.dtype == true_out_arr.dtype
+
+
+def test_binning_corner_cases():
+    # 0-dim
+    arr = np.array(1.0)
+    assert binning(arr, 2) == 1.0  # ok since there are no axes
+    with pytest.raises(ValueError):
+        binning(arr, [2])  # not ok, too many binning values
```
Comment (aringh) [inline odl/test/util/numerics_test.py:609]: I fail to see the difference between the two cases: why is the first one ok, but the second one is not?

File: odl/util/numerics.py [hunk 2]
```diff
@@ -847,6 +851,116 @@ def zscore(arr):
...
+def binning(arr, bin_size, reduction=np.sum):
+    """Bin an array by a factor.
+
+    Parameters
+    ----------
+    arr : `array-like`
+        The array that should be binned.
+    bin_size : positive int or sequence
+        Size or per-axis sizes of the bins.
+    reduction: callable
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/util/numerics.py:863]: Of course

File: odl/test/util/numerics_test.py (hunk 1 above)
Comment (kohr-h) [reply odl/test/util/numerics_test.py:609]: It's largely a consequence of the implementation. `binning(arr, scalar)` is equivalent to `binning(arr, [scalar] * arr.ndim)`, which is `binning(arr, [])` in the first case. It results in something like `np.sum(arr.reshape(()), axis=())`, which seems to work fine.
The other one is clearly a violation because `bin_sizes` is longer than `ndim`.

File: odl/util/numerics.py [hunk 3]
```diff
@@ -847,6 +851,116 @@ def zscore(arr):
...
+    d = arr.ndim
+    try:
+        bin_sizes = [int(bin_size)] * d
+    except TypeError:
+        bin_sizes = bin_size
+
+    if not all(b > 0 for b in bin_sizes):
+        raise ValueError('expected positive `bin_size`, got {}'
+                         ''.format(bin_size))
+    if len(bin_sizes) > d:
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/util/numerics.py:917]: Hm, that *might* work, but I haven't checked. If it works in a useful way, it may be worth keeping, but otherwise it'd be better to filter it out.

[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/util/numerics.py]: It doesn't work, now it's also disallowed

File: odl/test/util/numerics_test.py (hunk 1 above)
Comment (kohr-h) [reply odl/test/util/numerics_test.py:612]: I'd argue that this is consistent behavior: "reduce over all axes" just means "don't reduce" if there are no axes.
